File: WdtcCore/src/main/java/org/wdt/wdtc/download/fabric/FabricAPIVersionList.java
package org.wdt.wdtc.download.fabric;

import org.wdt.utils.gson.JSONArray;
import org.wdt.utils.gson.JSONObject;
import org.wdt.wdtc.download.infterface.VersionList;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.utils.PlatformUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FabricAPIVersionList implements VersionList {
    private static final String VersionListUrl = "https://api.modrinth.com/v2/project/P7dR8mSH/version";
    private final Launcher launcher;

    public FabricAPIVersionList(Launcher launcher) {
        this.launcher = launcher;
    }

    @Override
    public List<String> getVersionList() throws IOException {
        List<String> VersionList = new ArrayList<>();
        JSONArray VersionListArray = JSONArray.parseJSONArray(PlatformUtils.UrltoString(VersionListUrl));
        for (int i = 0; i < VersionListArray.size(); i++) {
            JSONObject VersionObject = VersionListArray.getJSONObject(i);
            JSONArray GameVersionsList = VersionObject.getJSONArray("game_versions");
            for (int j = 0; j < GameVersionsList.size(); j++) {
                if (launcher.getVersionNumber().equals(GameVersionsList.getString(j))) {
                    VersionList.add(VersionObject.getString("version_number"));
                }
            }
        }
        return VersionList;
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/fabric/FabricAPIVersionList.kt
package org.wdt.wdtc.core.impl.download.fabric

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.openapi.download.interfaces.VersionListInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asVersionsObjectSequence
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.utils.ioAsync
import org.wdt.wdtc.core.openapi.utils.toURL
import java.net.URL
import java.util.*

class FabricAPIVersionList(private val version: Version) : VersionListInterface {
	
	private val versionListUrl = "https://api.modrinth.com/v2/project/P7dR8mSH/version".toURL()
	
	private val list = ioAsync { versionListUrl.toStrings().parseJsonArray() }
	
	override suspend fun getVersionList(): VersionsObjectSequence {
		return list.await().asSequence<FabricAPIVersionsJsonObjectImpl>().filter {
			version.versionNumber == it.gameVersion.first()
		}.asVersionsObjectSequence()
	}
}


class FabricAPIVersionsJsonObjectImpl(
	@field:SerializedName("version_number")
	override val versionNumber: String,
	
	@field:SerializedName("date_published")
	val releaseTime: Date,
	
	@field:SerializedName("game_versions")
	val gameVersion: List<String>,
	
	@field:SerializedName("files")
	val files: List<FilesObject>,
	
	) : VersionsJsonObjectInterface {
	
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		val that = other as FabricAPIVersionsJsonObjectImpl
		return versionNumber == that.versionNumber
	}
	
	override fun hashCode(): Int {
		return Objects.hash(versionNumber)
	}
	
	
	class FilesObject(
		@field:SerializedName("url")
		val url: URL,
		@field:SerializedName("size")
		val size: Int,
		@field:SerializedName("filename")
		val name: String
	)
}
